exec/vector/src/main/java/org/apache/drill/exec/record/MaterializedField.java

```diff
@@ -49,39 +54,79 @@ private MaterializedField(String name, MajorType type, LinkedHashSet<Materialize
     this.children = children;
   }
 
+  private MaterializedField(String name, MajorType type, int size) {
+    this(name, type, new LinkedHashSet<>(size));
+  }
+
+  private <T> void copyFrom(Collection<T> source, Function<T, MaterializedField> transformation) {
+    Preconditions.checkState(children.isEmpty());
+    source.forEach(child -> children.add(transformation.apply(child)));
+  }
+
+  public static MaterializedField create(String name, MajorType type) {
+    return new MaterializedField(name, type, 0);
+  }
+
   public static MaterializedField create(SerializedField serField) {
-    LinkedHashSet<MaterializedField> children = new LinkedHashSet<>();
-    for (SerializedField sf : serField.getChildList()) {
-      children.add(MaterializedField.create(sf));
+    MaterializedField field = new MaterializedField(serField.getNamePart().getName(), serField.getMajorType(), serField.getChildCount());
+    if (OFFSETS_FIELD.equals(field)) {
+      return OFFSETS_FIELD;
     }
-    return new MaterializedField(serField.getNamePart().getName(), serField.getMajorType(), children);
+    field.copyFrom(serField.getChildList(), MaterializedField::create);
+    return field;
   }
 
-  /**
-   * Create and return a serialized field based on the current state.
-   */
-  public SerializedField getSerializedField() {
-    SerializedField.Builder serializedFieldBuilder = getAsBuilder();
-    for(MaterializedField childMaterializedField : getChildren()) {
-      serializedFieldBuilder.addChild(childMaterializedField.getSerializedField());
+  public MaterializedField copy() {
+    return copy(getName(), getType());
+  }
+
+  public MaterializedField copy(MajorType type) {
+    return copy(name, type);
+  }
+
+  public MaterializedField copy(String name) {
+    return copy(name, getType());
+  }
+
+  public MaterializedField copy(String name, final MajorType type) {
+    if (this == OFFSETS_FIELD) {
+      return this;
     }
-    return serializedFieldBuilder.build();
+    MaterializedField field = new MaterializedField(name, type, getChildren().size());
+    field.copyFrom(getChildren(), MaterializedField::copy);
+    return field;
   }
 
-  public SerializedField.Builder getAsBuilder() {
-    return SerializedField.newBuilder()
-        .setMajorType(type)
-        .setNamePart(NamePart.newBuilder().setName(name).build());
+  public String getName() {
+    return name;
+  }
+
+  public MajorType getType() {
+    return type;
   }
 
   public Collection<MaterializedField> getChildren() {
-    return new ArrayList<>(children);
+    return children;
+  }
+
+  public int getWidth() {
+    return type.getWidth();
+  }
```

Comment:
Width is deprecated. We use precision now where width used to be used.

Reply:
`getWidth()` is still used.

Reply:
It should not be. @arina-ielchiieva did a PR a year ago that deprecated width. References to width represent bugs.

Reply:
I don't know why it was deprecated instead of being completely removed. If it needs to be removed, somebody has to open another JIRA/PR.